### PR TITLE
Add Azure support for ranged read blob operations

### PR DESCRIPTION
**Pull request by tlrx**
This pull request adds support for ranged read blob operations to the repository azure plugin. It adds the necessary plumbing down to the Azure SDK with additionnal unit tests and also adds a QA test for searchable snapshots.

Relates #50999

### CODE REVIEW

File: plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java [hunk 1]
```diff
@@ -198,6 +214,57 @@ public void testReadBlobWithRetries() throws Exception {
...
+        final byte[] bytes = randomBlobContent();
+        httpServer.createContext("/container/read_range_blob_max_retries", exchange -> {
+            Streams.readFully(exchange.getRequestBody());
+            if ("HEAD".equals(exchange.getRequestMethod())) {
+                if (countDownHead.countDown()) {
+                    exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
+                    exchange.getResponseHeaders().add("x-ms-blob-content-length", String.valueOf(bytes.length));
+                    exchange.getResponseHeaders().add("x-ms-blob-type", "blockblob");
+                    exchange.sendResponseHeaders(RestStatus.OK.getStatus(), -1);
+                    exchange.close();
```
Comment (original-brownbear) [inline plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java]: Maybe `try { ... } finally { exchange.close()}` around the whole handler so we don't get weird exceptions from the REST mock if an assertion trips?

**Review by original-brownbear — APPROVED** [review]
LGTM :)

File: plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java (hunk 1 above)
Comment (tlrx) [reply plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java]: Sure

File: plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java [hunk 2]
```diff
@@ -128,7 +135,7 @@ private BlobContainer createBlobContainer(final int maxRetries) {
         final AzureStorageService service = new AzureStorageService(clientSettings.build()) {
             @Override
             RetryPolicyFactory createRetryPolicy(final AzureStorageSettings azureStorageSettings) {
-                return new RetryExponentialRetry(1, 100, 500, azureStorageSettings.getMaxRetries());
+                return new RetryExponentialRetry(1, 10, 100, azureStorageSettings.getMaxRetries());
```
[the remark this answers is not in the document]
Comment (tlrx) [reply plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java]: While looking at this class I found them a bit high, specially when a high number of retries (5) is picked up so I lowered them. If that causes any issue on CI I'll restore the previous values.

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by tlrx** [comment]
@elasticmachine update branch

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/1 (unrelated failure)
@elasticmachine run elasticsearch-ci/2 (IML [related failure](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-2/20425/))

**Comment by tlrx** [comment]
Thanks a lot Armin!

**Pull request merged**